includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -291,6 +291,8 @@ private function sanitize_element( DOMElement $element ) {
 						$result
 					);
 				}
+			} elseif ( $this_child instanceof DOMProcessingInstruction ) {
+				$this->remove_invalid_child( $this_child, [ 'code' => 'invalid_processing_instruction' ] );
```

Comment:
Please also check the other DOM node types that are not allowed. For example, a `DOMCdataSection` is not allowed in AMP:

![image](https://user-images.githubusercontent.com/134745/67037521-8663da00-f0d2-11e9-95d4-2335c8c52b96.png)

Try adding checks for all of the various `dom` classes that can be generated by the HTML DOM parser: https://www.php.net/manual/en/book.dom.php

Perhaps `DOMNotation` (e.g. `<!NOTATION …>`), `DOMEntity` (e.g. `<!ENTITY …>`), and `DOMEntityReference` (e.g. `&foo;`)? Not every DOM class (or [node type](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType)) is possible to be instantiated from parsing, but please check them for good measure.

Reply:
According to the [DOM standard](https://dom.spec.whatwg.org/#node-trees) the possible nodes are `Document`, `DocumentType`, `DocumentFragment`, `Element`, `Text`, `ProcessingInstruction`, and `Comment`, which I think is all we need to consider.

The only invalid node then that should not be allowed in AMP would be `ProcessingInstruction`.

Reply:
OK, I wasn't able to cause the addition of a CDATA to result in an AMP validation error; the HTML parser actually just drops them from the document it seems. So yes, it appears we don't need to worry about them.

Reply:
> the HTML parser actually just drops them from the document 

Yes, that's what I found out as well. Reduces the work for us :smile: .